internal/engine/upper.go

```diff
@@ -139,6 +147,22 @@ func (u Upper) CreateServices(ctx context.Context, services []model.Service, wat
 	return nil
 }
 
+func (u Upper) getContainersForBuildStates(ctx context.Context,
```

Comment:
the api signature of this method sounds like you went back and forth on a couple different directions

it uses lots of pointers (which seems to indicate that you're modifying things in place)
it's also called getContainers (which seems to indicate that you're not modifying things in place)
and the actual implementation does both (mutates things in-place and returns things)

let's change this data structure back to a `map[model.ServiceName]BuildState`, and rename the method to `populateContainersForBuildStates`. happy to pair on it if you're confused about mutability and pointers.